pkg/controller/garbagecollector/graph_builder.go

```diff
@@ -170,37 +219,71 @@ func (gb *GraphBuilder) controllerFor(resource schema.GroupVersionResource, kind
 	return monitor, nil
 }
 
-func (gb *GraphBuilder) monitorsForResources(resources map[schema.GroupVersionResource]struct{}) error {
+// syncMonitors starts or stops monitors as necessary. It will return any error
+// encountered, but will make an attempt to start a monitor for each resource
+// instead of immediately exiting on an error. It may be called before or after Run.
+func (gb *GraphBuilder) syncMonitors(resources map[schema.GroupVersionResource]struct{}) error {
+	gb.monitorLock.Lock()
+	defer gb.monitorLock.Unlock()
+
+	toRemove := gb.monitors
+	if toRemove == nil {
+		toRemove = monitors{}
+	}
+	current := monitors{}
+	errs := []error{}
 	for resource := range resources {
 		if _, ok := ignoredResources[resource]; ok {
-			glog.V(5).Infof("ignore resource %#v", resource)
+			continue
+		}
+		if m, ok := toRemove[resource]; ok {
+			current[resource] = m
+			delete(toRemove, resource)
 			continue
 		}
 		kind, err := gb.restMapper.KindFor(resource)
 		if err != nil {
-			nonCoreMsg := fmt.Sprintf(nonCoreMessage, resource)
-			utilruntime.HandleError(fmt.Errorf("%v. %s", err, nonCoreMsg))
+			errs = append(errs, fmt.Errorf("couldn't look up resource %q: %v", resource, err))
 			continue
 		}
-		monitor, err := gb.controllerFor(resource, kind)
+		m, err := gb.controllerFor(resource, kind)
 		if err != nil {
-			return err
+			errs = append(errs, fmt.Errorf("couldn't start monitor for resource %q: %v", resource, err))
+			continue
 		}
-		gb.monitors = append(gb.monitors, monitor)
+		monitor := &monitorSingle{monitor: m}
+		if gb.globalStop != nil {
+			monitor.Run(gb.globalStop)
+		}
+		current[resource] = monitor
+	}
+	gb.monitors = current
+
+	for _, monitor := range toRemove {
+		if monitor.stopper != nil {
+			close(monitor.stopper.localStop)
+		}
+		// TODO: remove all resources of this type from the graph?
 	}
-	return nil
+
+	// NewAggregate returns nil if errs is 0-length
+	return utilerrors.NewAggregate(errs)
 }
 
 func (gb *GraphBuilder) HasSynced() bool {
 	for _, monitor := range gb.monitors {
```

Comment:
this misses a lock. Or at the very least it should be documented that it assumes it being taken.